FREEWARE/v40/snake/snake4.6
.TH SNAKE4 6 "11 October 1996" \" -*- nroff -*-

.SH NAME
snake4, snake4scores \- Game starring a fruit-eating snake.

.SH SYNOPSIS
.B snake4
[\-hHV] [\-M file]
[\-\-help] [\-\-highscores] [\-\-version]
[\-\-merge\-highscores=file]

.B snake4scores

.SH DESCRIPTION
The
.B snake4
program is my fourth implementation of the snake game, written for
Unix and X11. Previous versions were written for MS-DOS.

Move the snake around the screen and eat food. The snake grows in
length as you eat. Avoid hitting the fence, eating lethal mushrooms
and rotten food, and biting your own tail.

Now and then a slim-pill will bounce around. If you eat this, you will
get shorter, and you will gain bonus points for every length you
loose. The bonus is also given when a round is over.

When food starts blinking, it's about to rot. Rotten food is
poisonous, and thus uneatable. As long as it blinks, it can still be
eaten.

Oh, yes - I almost forgot: You may be visited by the evil headbanger.
Take my advice and run away from him before he hits you in the head!
He won't stay long.

The keys used to control the game are given at startup.

.B snake4
features a site-wide highscore file. The file contains the top 200
users. Each user may appear once only in the highscore table.

For technical reasons,
.B snake4
connects to the X11 server before parsing options. This makes it
impossible to eg. browse the highscore table using -H if no server is
available.
To be able to see the scores in such a case,
.B snake4scores
produces the same output as
.B snake4
with option -H, but does no attempt on connecting to the X11 server.

.SS OPTIONS
.TP
.I "-h, \-\-help"
Print a usage message on standard output, and exit successfully.

.TP
.I "\-H, \-\-highscores"
Dump the highscore list to standard output, and exit successfully.
Starting
.B snake4
with the name
.B snake4score
does the same thing, without first connecting to the X11 server.  Note
that this will possibly print more entries than the 20 or so that will
fit within the game startup screen.

.TP
.I "-V, \-\-version"
Print version information on standard output, then exit successfully.

.TP
.I "\-M, \-\-merge\-highscores=file"
Merge separate highscore file
.I file
into the default highscore file, and exit successfully. Displays the
new highscore table. This option is for the game owner or root only.

.SH X11 OPTIONS
In addition to the above mentioned options,
.B snake4
accepts the standard X11 toolkit options.

.SH AUTHOR
Sverre H. Huseby <sverrehu@ifi.uio.no>.


FREEWARE/v40/snake/snake4.c
/* $Id: snake4.c,v 1.5 1996/10/09 16:11:48 sverrehu Exp $ */
/**************************************************************************
 *
 *  FILE            snake4.c
 *  MODULE OF       snake4 - game of snake eating fruit
 *
 *  DESCRIPTION     
 *
 *  WRITTEN BY      Sverre H. Huseby <sverrehu@ifi.uio.no>
 *
 **************************************************************************/

#include <stdlib.h>
#include <string.h>

#ifndef VMS
#include <shhmsg.h>
#include <shhopt.h>
#else
#define COMPILED_DATE "1998"
#define COMPILED_BY "DEC C"
#endif

#include "win.h"
#include "game.h"
#include "score.h"
#include "suid.h"

/**************************************************************************
 *                                                                        *
 *                   P R I V A T E    F U N C T I O N S                   *
 *                                                                        *
 **************************************************************************/

static void
version(void)
{
    printf(
      "snake4 " VERSION ", by Sverre H. Huseby "
      "(compiled " COMPILED_DATE " by " COMPILED_BY ")\n"
    );
    exit(0);
}

static void
usage(void)
{
#ifndef VMS
    printf(
      "usage: %s [options]\n"
      "\n"
      "  -h, --help                    display this help and exit\n"
      "  -H, --highscores              show highscores and exit\n"
      "  -M, --merge-highscores=FILE   merge highscores from FILE and exit\n"
      "  -V, --version                 output version information and exit\n"
      "\n"
      "In addition, `normal' X11 options are allowed.\n",
      msgGetName()
    );
#else
    printf(
      "usage: snake4 [options]\n"
      "\n"
      "  -h, --help                    display this help and exit\n"
      "  -H, --highscores              show highscores and exit\n"
      "  -M, --merge-highscores=FILE   merge highscores from FILE and exit\n"
      "  -V, --version                 output version information and exit\n"
      "\n"
      "In addition, `normal' X11 options are allowed.\n");
#endif
    exit(0);
}

static void
highscores(void)
{
    scoreDumpHighscores();
    exit(0);
}

static void
merge(char *file)
{
    scoreMergeScoreFile(file);
    scoreDumpHighscores();
    exit(0);
}



/**************************************************************************
 *                                                                        *
 *                    P U B L I C    F U N C T I O N S                    *
 *                                                                        *
 **************************************************************************/

int
main(int argc, char *argv[])
{
#ifndef VMS
    optStruct opt[] = {
      /* short long           type        var/func    special       */
        { 'h', "help",        OPT_FLAG,   usage,      OPT_CALLFUNC },
        { 'H', "highscores",  OPT_FLAG,   highscores, OPT_CALLFUNC },
        { 'M', "merge-highscores",
                              OPT_STRING, merge,      OPT_CALLFUNC },
        { 'V', "version",     OPT_FLAG,   version,    OPT_CALLFUNC },
        { 0, 0, OPT_END, 0, 0 }  /* no more options */
    };
    msgSetName(argv[0]);
    suidInit();

    /* to allow showing highscores before connecting to X11 server */
    if (strcmp(msgGetName(), "snake4scores") == 0)
	highscores(); /* exits the program */
#endif

    winParseOptions(&argc, argv);
#ifndef VMS
    optParseOptions(&argc, argv, opt, 0);
#endif

    gameInit();

    winMainLoop();

    gameFinish();

    return 0;
}
